### PR TITLE
Multiple inheritance collection serialization

### PR DESCRIPTION
**Pull request by bwegrzyn**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1025
| License       | MIT
| Doc PR        | 

As I am having the same issues as #1025, I figured I'd contribute a failing test case to help move things along on PR #1046. 

@nik-lampe @dunglas @soyuka Can you take a look at this test case and make sure it matches your failing use case?

### CODE REVIEW

**Comment by bwegrzyn** [comment]
Curiosity got the better of me, and I found a super simple fix while debugging the issue. It fixed all my use cases and all behat tests, including the ones I added, are green.

However, there is one failing phpunit tests. It looks like it just might have to be updated, but I'm not sure. ping @dunglas @soyuka can you take a look?

File: src/JsonLd/Serializer/ItemNormalizer.php [hunk 1]
```diff
@@ -66,6 +66,9 @@ public function normalize($object, $format = null, array $context = [])
         $resourceMetadata = $this->resourceMetadataFactory->create($resourceClass);
         $data = $this->addJsonLdContext($this->contextBuilder, $resourceClass, $context);
 
+        // Use resolved resource class instead of given resource class to support multiple inheritance child types
+        $context['resource_class'] = $resourceClass;
```
Comment (soyuka) [inline src/JsonLd/Serializer/ItemNormalizer.php]: Is this true for classes without inheritance?

**Comment by soyuka** [comment]
This fixes test, but it's kinda weird that we now call `getResourceClass` twice with different parameters.

```diff
index e474a5b..fdc57b5 100644
--- a/tests/Hydra/Serializer/ItemNormalizerTest.php
+++ b/tests/Hydra/Serializer/ItemNormalizerTest.php
@@ -102,6 +102,7 @@ class ItemNormalizerTest extends \PHPUnit_Framework_TestCase

         $resourceClassResolverProphecy = $this->prophesize(ResourceClassResolverInterface::class);
         $resourceClassResolverProphecy->getResourceClass($dummy, null, true)->willReturn(Dummy::class)->shouldBeCalled();
+        $resourceClassResolverProphecy->getResourceClass($dummy, Dummy::class, true)->willReturn(Dummy::class)->shouldBeCalled();

         $serializerProphecy = $this->prophesize(SerializerInterface::class);
         $serializerProphecy->willImplement(NormalizerInterface::class);
```

**Comment by bwegrzyn** [comment]
Yes, I've noticed that and that's how I fixed the unit test as well, but I have not committed it yet.

I agree that it is a little weird, but I'm not sure how to go about fixing it another way since ItemNormalizer->normalize is where we first are able to figure out the proper resource class to normalize. Any ideas?

**Comment by soyuka** [comment]
I think modifying the test is fine. Indeed, we force the `context['resource_class']` on the next `normalize` passes.

**Comment by soyuka** [comment]
@nik-lampe does this patch fixes your issue?

**Comment by nik-lampe** [comment]
Yes, looks pretty good so far. It is a similar change like I made in my code, but much better.

**Comment by nik-lampe** [comment]
I just came across one issue, that was solved with my dirty hacked fix but came back when using the fix from @bwegrzyn. Can't pin it down right now, but I'm on it

**Comment by bwegrzyn** [comment]
@nik-lampe Do you have a failing test case for it or an example? I tried to cover my changes with all the use cases I could think of, but I may have missed some.

**Comment by nik-lampe** [comment]
@bwegrzyn I took a quick look and the issue I described in #1043 might be a case. At least it fails with your changes but works with mine

**Comment by bwegrzyn** [comment]
@nik-lampe It looks like that issue is separate from this issue and not related to inheritance. I think this PR should be merged if it is approved, and the other issue should be taken care of in a new PR. 

I'll take a look at #1043 and see if I can help you with it.

**Comment by nik-lampe** [comment]
That's true, it has nothing to do with inheritance. But it results from having the wrong resource class in the context when trying to access the property metadata. I agree with you, that it should be taken care of in a new PR.

**Comment by bwegrzyn** [comment]
@dunglas I'm having some trouble figuring out how to do that properly. Can you help me out?

**Comment by soyuka** [comment]
Thanks @bwegrzyn !

**Pull request merged**